File: surface/widgets/compose/components/widgets/concepts/governance-execution/GuardStatusPanel.kt
package com.clef.surface.widgets.concepts.governanceexecution

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.itemsIndexed
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// --- State machine ---

enum class GuardStatusPanelState { Idle, GuardSelected }

sealed class GuardStatusPanelEvent {
    data class SelectGuard(val id: String?) : GuardStatusPanelEvent()
    object GuardTrip : GuardStatusPanelEvent()
    object Deselect : GuardStatusPanelEvent()
}

fun guardStatusPanelReduce(state: GuardStatusPanelState, event: GuardStatusPanelEvent): GuardStatusPanelState = when (state) {
    GuardStatusPanelState.Idle -> when (event) {
        is GuardStatusPanelEvent.SelectGuard -> GuardStatusPanelState.GuardSelected
        is GuardStatusPanelEvent.GuardTrip -> GuardStatusPanelState.Idle
        else -> state
    }
    GuardStatusPanelState.GuardSelected -> when (event) {
        is GuardStatusPanelEvent.Deselect -> GuardStatusPanelState.Idle
        else -> state
    }
}

// --- Types ---

enum class GuardStatus(val icon: String, val label: String) {
    Passing("\u2713", "Passing"),
    Failing("\u2717", "Failing"),
    Pending("\u23F3", "Pending"),
    Bypassed("\u2298", "Bypassed")
}

data class Guard(
    val id: String? = null,
    val name: String,
    val description: String,
    val status: GuardStatus,
    val lastChecked: String? = null
)

private fun deriveOverallStatus(guards: List<Guard>): String {
    if (guards.isEmpty()) return "all-passing"
    if (guards.any { it.status == GuardStatus.Failing }) return "has-failing"
    if (guards.any { it.status == GuardStatus.Pending }) return "has-pending"
    return "all-passing"
}

@Composable
fun GuardStatusPanel(
    guards: List<Guard>,
    modifier: Modifier = Modifier,
    onGuardSelect: (String?) -> Unit = {}
) {
    var state by remember { mutableStateOf(GuardStatusPanelState.Idle) }
    var selectedGuardId by remember { mutableStateOf<String?>(null) }

    val overallStatus = remember(guards) { deriveOverallStatus(guards) }
    val hasBlocking = overallStatus == "has-failing"
    val selectedGuard = remember(selectedGuardId, guards) {
        selectedGuardId?.let { id -> guards.find { (it.id ?: it.name) == id } }
    }

    fun handleSelect(guardId: String) {
        val nextId = if (guardId == selectedGuardId) null else guardId
        selectedGuardId = nextId
        onGuardSelect(nextId)
        state = if (nextId != null) GuardStatusPanelState.GuardSelected else GuardStatusPanelState.Idle
    }

    Column(modifier = modifier.semantics { contentDescription = "Pre-execution guards" }) {
        // Header
        Row(
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.padding(horizontal = 12.dp, vertical = 8.dp)
        ) {
            Text("Pre-execution Guards", style = MaterialTheme.typography.titleSmall, fontWeight = FontWeight.Bold)
            Spacer(Modifier.weight(1f))
            val statusText = when (overallStatus) {
                "all-passing" -> "\u2713 All passing"
                "has-failing" -> "\u2717 Blocked"
                "has-pending" -> "\u23F3 Pending"
                else -> "Unknown"
            }
            val statusColor = when (overallStatus) {
                "all-passing" -> MaterialTheme.colorScheme.primary
                "has-failing" -> MaterialTheme.colorScheme.error
                else -> MaterialTheme.colorScheme.onSurfaceVariant
            }
            Text(statusText, fontSize = 13.sp, color = statusColor, fontWeight = FontWeight.Medium)
        }

        // Blocking banner
        if (hasBlocking) {
            Surface(
                color = MaterialTheme.colorScheme.errorContainer,
                modifier = Modifier.fillMaxWidth().padding(horizontal = 12.dp, vertical = 4.dp)
            ) {
                Text(
                    "Execution blocked by failing guards",
                    color = MaterialTheme.colorScheme.onErrorContainer,
                    fontSize = 13.sp,
                    modifier = Modifier.padding(8.dp)
                )
            }
        }

        HorizontalDivider()

        // Guard list
        LazyColumn(modifier = Modifier.weight(1f)) {
            itemsIndexed(guards, key = { _, g -> g.id ?: g.name }) { _, guard ->
                val guardId = guard.id ?: guard.name
                val isSelected = selectedGuardId == guardId

                val statusColor = when (guard.status) {
                    GuardStatus.Passing -> MaterialTheme.colorScheme.primary
                    GuardStatus.Failing -> MaterialTheme.colorScheme.error
                    GuardStatus.Pending -> MaterialTheme.colorScheme.onSurfaceVariant
                    GuardStatus.Bypassed -> MaterialTheme.colorScheme.outline
                }

                Column(
                    modifier = Modifier
                        .fillMaxWidth()
                        .clickable { handleSelect(guardId) }
                        .then(
                            if (isSelected) Modifier.background(MaterialTheme.colorScheme.primaryContainer.copy(alpha = 0.2f))
                            else Modifier
                        )
                        .padding(horizontal = 12.dp, vertical = 8.dp)
                ) {
                    Row(
                        verticalAlignment = Alignment.CenterVertically,
                        horizontalArrangement = Arrangement.spacedBy(8.dp)
                    ) {
                        Text(guard.status.icon, fontSize = 16.sp, color = statusColor)
                        Text(guard.name, fontWeight = FontWeight.Medium, fontSize = 14.sp, modifier = Modifier.weight(1f))
                        Text(guard.status.label, fontSize = 12.sp, color = statusColor)
                    }

                    // Expanded detail when selected
                    if (isSelected) {
                        Column(modifier = Modifier.padding(start = 24.dp, top = 4.dp)) {
                            Text(guard.description, fontSize = 13.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
                            guard.lastChecked?.let {
                                Text("Last checked: $it", fontSize = 11.sp, color = MaterialTheme.colorScheme.onSurfaceVariant)
                            }
                        }
                    }
                }

                HorizontalDivider()
            }
        }
    }
}

File: surface/widgets/watchkit/components/widgets/concepts/governance-execution/GuardStatusPanel.swift
import SwiftUI

// State machine: idle | guardSelected
enum GuardStatusPanelWatchState {
    case idle
    case guardSelected
}

enum GuardStatusPanelWatchEvent {
    case selectGuard(String)
    case guardTrip
    case deselect
}

func guardStatusPanelWatchReduce(_ state: GuardStatusPanelWatchState, _ event: GuardStatusPanelWatchEvent) -> GuardStatusPanelWatchState {
    switch state {
    case .idle:
        switch event {
        case .selectGuard: return .guardSelected
        case .guardTrip: return .idle
        default: return state
        }
    case .guardSelected:
        if case .deselect = event { return .idle }
        return state
    }
}

struct GuardData: Identifiable {
    var id: String { name }
    let name: String
    let description: String
    let status: String // "passing", "failing", "pending", "bypassed"
    var lastChecked: String? = nil
}

struct GuardStatusPanelWatchView: View {
    let guards: [GuardData]
    let executionStatus: String
    var onGuardSelect: ((GuardData) -> Void)? = nil

    @State private var state: GuardStatusPanelWatchState = .idle
    @State private var selectedGuardName: String? = nil

    private var passingCount: Int { guards.filter { $0.status == "passing" }.count }
    private var hasBlocking: Bool { guards.contains { $0.status == "failing" } }

    private func statusIcon(_ status: String) -> String {
        switch status {
        case "passing": return "\u{2713}"
        case "failing": return "\u{2717}"
        case "pending": return "\u{23F3}"
        case "bypassed": return "\u{2298}"
        default: return "\u{2022}"
        }
    }

    private func statusColor(_ status: String) -> Color {
        switch status {
        case "passing": return .green
        case "failing": return .red
        case "pending": return .orange
        case "bypassed": return .secondary
        default: return .primary
        }
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 4) {
                Text("Guards")
                    .font(.caption2).fontWeight(.bold)
                Text("\(passingCount)/\(guards.count) passing")
                    .font(.system(size: 9)).foregroundColor(.secondary)

                if hasBlocking {
                    HStack(spacing: 4) {
                        Image(systemName: "exclamationmark.triangle.fill")
                            .font(.system(size: 9)).foregroundColor(.red)
                        Text("Blocked")
                            .font(.system(size: 9)).foregroundColor(.red)
                    }
                }

                Divider()

                ForEach(guards) { guard_ in
                    let isSelected = selectedGuardName == guard_.name
                    Button {
                        if isSelected {
                            selectedGuardName = nil
                            state = guardStatusPanelWatchReduce(state, .deselect)
                        } else {
                            selectedGuardName = guard_.name
                            state = guardStatusPanelWatchReduce(state, .selectGuard(guard_.name))
                            onGuardSelect?(guard_)
                        }
                    } label: {
                        VStack(alignment: .leading, spacing: 2) {
                            HStack(spacing: 4) {
                                Text(statusIcon(guard_.status))
                                    .foregroundColor(statusColor(guard_.status))
                                    .font(.system(size: 10))
                                Text(guard_.name)
                                    .font(.caption2)
                                    .lineLimit(1)
                                Spacer()
                                Text(guard_.status.capitalized)
                                    .font(.system(size: 8))
                                    .foregroundColor(statusColor(guard_.status))
                            }
                            if isSelected {
                                Text(guard_.description)
                                    .font(.system(size: 8))
                                    .foregroundColor(.secondary)
                                if let lastChecked = guard_.lastChecked {
                                    Text("Checked: \(lastChecked)")
                                        .font(.system(size: 7))
                                        .foregroundColor(.secondary)
                                }
                            }
                        }
                        .padding(.vertical, 2)
                        .background(isSelected ? Color.blue.opacity(0.1) : Color.clear)
                        .cornerRadius(3)
                    }
                    .buttonStyle(.plain)
                }
            }
            .padding(.horizontal, 4)
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel("Pre-execution guards: \(passingCount) of \(guards.count) passing")
    }
}
